### PR TITLE
Add support for JWKS from remote location

### PR DESCRIPTION
**Pull request by ArneLimburg**
@rmannibucau This patch implements support for JWKS i.e. for the use with Keycloak or Azure AD.

### CODE REVIEW

**Comment by rmannibucau** [comment]
@ArneLimburg no blocker for me but a few open points - see it the way I would have done it so I'd love to ensure it got reviewed before we integrate the code:

* do we want to support more than RSA keys (https://github.com/yupiik/fusion/blob/master/fusion-jwt/src/main/java/io/yupiik/fusion/jwt/JwtValidatorFactory.java#L91 can be a sample)
* do we want to load it at startup - postconstruct - and use it at need at runtime - loadKey - (ie https://github.com/apache/geronimo-jwt-auth/pull/7/files#diff-6cae8d1e2c724e267297472eb5909c6ecd132a7447d1af4924cb65de04940c73R148 becomes a HttpClient with a pool of 1 thread doing a sendAsync, storing the completionstage and blocking on it at runtime to ensure keys are there - since sadly API is not designed to be reactive)?
* we likely want to refresh the keys from time to time to avoid to force a restart of the app to handle key rotation (in particular when it is everyday for ex)
* I don't see the filtering on "sig", is it intended - maybe I missed it but don't think the spec supports enc yet?

Wdyt?

**Comment by ArneLimburg** [comment]
- Supporting an EC key should be no problem with your sample, will implement
- Storing the CompletionStage is a good idea, even though it will help only for the initial call. When we have a cache miss later it probably is because of a key rotation and then we have to block anyway.
- The best solution for that one would probably be a LRUMap. Since we don't have one I suggest to clear the map before putting new keys in. Wdyt?
- you are right, we'll implement the filtering

**Comment by rmannibucau** [comment]
> When we have a cache miss later it probably is because of a key rotation and then we have to block anyway.

Not really cause we refresh it in background. How I see things is a scheduled executor service following the manager lifecycle (postconstruct/predestroy) and the `refresh` of  submit(() -> refresh()) wrapped to be a volatile `CompletionFuture` - initialized in the postconstruct.
This way we always do `getNow(null)`. The promise is only needed during the init - at runtime we use it as a volatile field - but shouldn't be a big deal. Idea is to limit the blocking to almost never.

> LRUMap

I'm not sure, the rotation is not about the order since it uses identifier, we can store the remote identifiers, when we refresh compare to the one in the "soon new" map and remove the ones no more there only, sounds sufficient to me.

Does it make sense?

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: > return keys.stream()
>     .map(JsonValue::asJsonObject)
>     .map(JWK::new)
>     .filter(it -> Objects.equals("sig", it.getUse()) or it.getUse() == null /*i consider null as ok but fine to ignore it too for me*/) // no need of optional there plus the lazy creation is always done
>     .collect(toList());

;)

**Comment by ArneLimburg** [comment]
@rmannibucau I am still not sure, how to handle the key rotation case properly (i.e. how to handle the reload of keys).
For now I would reload the keyset when no key is present for the requested kid. But then we have to block or not?

**Comment by rmannibucau** [comment]
@ArneLimburg let me try to put some pseudo code on github:

```java
// volatile Set<String> remoteKeyIds
// volatile CompletionFuture<?> keyLoading
// ScheduledExecutorService backgroundThread
// HttpClient httpClient

@PostConstruct private void init() {
    readFromConfUri(...).ifPresent(uri -> {
        backgroundThread = Executors.newSingleScheduledExecutorService(new Threadfactory() { /* name + ThisClass.class.getClassLoader() on the thread */ });
        httpClient = HttpClient.newBuilder().executor(backgroundThread).....build();
        keyLoading = reloadRemoteKeys();
        readFromConfRefreshIntervalMs(...).filter(i -> i > 0).ifPresent(i -> backgroundThread.scheduledAtFixedRate(this::reloadRemoteKeys, i, MILLISECONDS, i, MILLISECONDS);
    });
}

private Key getKey(String id) {
  if (!keyLoading.isDone()) { keyLoading.get(1, MINUTE); } // only give it a chance for the initial loading, then at runtime it is always true even for rotations
  // as now
}

private void reloadRemoteKeys() {
    keyLoading = httpClient.sendAsync(....).thenAccept(res -> {
        check200(res);
        final var keys = parse(res.body());
        final var newRemoteKeyIds = ids(keys);
        final var noMorePresent = diffRemoved(remoteKeyIds, newRemoteKeyIds);
        noMorePresent.forEach(keyMapping::removeKey);
        keyMapping.putAll(keys);
        remoteKeyIds = newRemoteKeyIds;
    });
}

@PreDestroy private void destroy() {
  // if != null ... on all statement
  keyLoading.get(1, MINUTES); // if pending give it a chance to finish
  backgroundThread.shutdownNow(); + await (1mn max and then give up?)
  if (httpClient instance AutoCloseable) { close(); } // java >= 21
}
```

**Comment by ArneLimburg** [comment]
OK, you are talking about a fixed reload interval. I was thinking more of a dynamic reloading when key rotation is detected (i.e. because the kid is not contained in the cache). Don't you think reloading on a key miss is enough (also it would be blocking for that request)?

**Comment by rmannibucau** [comment]
Doing it on a key miss will add the reload latency so network latency - depends if the oauth2 server is in the same vpc or not - and cert reloading - depends how much you have and #cpus.
When it will happen all calls will block which kind of misbehave for me.

If the thread bothers you we can fire an event with the lambda to reload the state - event.refresh()? - and let the integrator handle it with its own threads optionally if well documented.

**Comment by ArneLimburg** [comment]
@rmannibucau Would you mind to take another look and merge, if ok?

Maybe directly doing a release?

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: private?

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java:113]: guess minimum is a `protected void customize(HttpClient.Builder)` method - to be transparent I can't replace by custom code fillung the key map with that cause of missing sslContext+sslParameters setup, not asking to include it by default but enable it - guess the protected method is the fastest but I'm fine with an event too.

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java:148]: can be worth to be protected too, some servers are not that friendly (don't hit, not my choice ;))

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: `.thenApplyAsync(...., httpClient.executor().orElseThrow())` otherwise it is ran in commonPool which is something we don't want nor we control there.

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: > keyMapping = parseKeys(result).stream().collect(toMap(JWK::kid, JWK::toPemKey));

side note: should we tolerate kid-less keys if single and use the default id?

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: > try (final JsonReader reader = readerfactory.createReader(new StringReader(keyResponse.body()))) {...}

otherwise you leak memory (jsonp)

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: tip: filter should be before map generally speaking even if not critical

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: + await (1mn can be hardcoded since it should be almost immediate - one call duration) to try our best to not leak a thread before quitting this method (the class being the thread owner)

side note: the cancel should be done before the await

File: src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/JwksServer.java
Comment (rmannibucau) [inline src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/JwksServer.java:10]: why not `HttpServer`? - not a blocker more curious

File: src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapperTest.java
Comment (rmannibucau) [inline src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapperTest.java]: where is it cleaned?

tip: inject it in the war as a mp properties file

side note: true for all tests

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: volatile (not critical at all, just want to ensure we thought about it and use it or not deliberatly)

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: store the interval in a local var to avoid to call it N times

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: if interval is null or < 0 just drop the client (`httpClient=null`)

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: `ThreadFactory` to name the thread (kill -3 friendly ;))

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: not needed, just set `reloadJwksRequest` in previous line

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [inline src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: split `InterruptedException` and `Thread.currentThread().interrupt()` there ?

catch of `ExecutionException` should `throw e.getCause() instanceof RuntimeException ? (RuntimeException) e.getCause() : new IllegalStateException(e.getCause());` probably

**Review by rmannibucau — CHANGES_REQUESTED** [review]
Some requests (like enabling to tune the http client since this does not enable to use this feature in dev for example - even if not "great" env, it is common) + some open questions to ensure we didn't miss things but we are 100% in the expected pattern so think we can make it in pretty soon

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: If we set it in previous line, the code will block, every time a reload takes place. This way it will just take the old key set or am I wrong?

Comment (rmannibucau) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: ok, let me detail my view to ensure we both get the same picture - think we both see it from a different side:

* overall I think we should always use the "live" mapping and ignore the `.get()` -> seems we agree on that
* however there is a case we want the `.get()` to happen and potentially block, ie when the first call is not finished yet, there the first call in the constructor must be `this.reloadRemoteKeys = ongoingRequest // from reloadRemoteKeys()` otherwise this first hit can be missed and the call can fail whereas it should pass
* we don't want `.get()` to be called in postconstruct but lazily - hoping it is not called

My proposal would be:
1. in postconstruct set the promise to the actual promise and not post execution (thenApply) as proposed
2. in `loadKey` dont test `reloadJwksRequest!=null` but `if (reloadJwksRequest!=null&&!reloadJwksRequest.isDone())`
3. don't refresh `this.reloadJwksRequest` so we always test the postconstruct value
4. in postconstruct add `.thenRun(() -> this.reloadJwksRequest = null)` to avoid the `isDone()` call at runtime (optim but optional if you don't like that)

side note: reloadJwksRequest could then be renamed `initialJwksPromise` or alike to show the intent.

hope it makes sense

Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: Yes, we agree on the first point.

My intention was to alway use the `.get`, but ensure, that it only blocks in the initial case. That written I removed the null-check in `tryLoad`. `thenApply` and `thenRun` will run in the thread that handles the HTTP-response, so it should be fine.

Comment (rmannibucau) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: Hmm, sorry I'm not sure I understood: you will rework it as proposed cause as of today the theRun does not do the intent nor runs in the http response thread - it uses commonpool on hotspot and a specific cached thread pool on openj9/IBM semuru.

File: src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/JwksServer.java
Comment (ArneLimburg) [reply src/test/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/JwksServer.java:10]: I guess, it was, because we just had the code lying around.

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: We need the client for the initial request in `reloadRemoteKeys()`

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: OK, I guess I am not so into the Executor/ThreadFactory stuff. How should this look exactly and why would that help?

Maybe we should put the creation of the Executor in a separate protected method anyway, to allow overriding and using other Executor, wdyt?

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: this is ok you can wire a thenApply to `if (client instanceof AutoCloseable) { ((AutoCloseable) client).close(); } client=null;` for example or any other refactor (like passing the client to the reload method).

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (rmannibucau) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: ok for the protected method for me

for the thread factory just pass a param: `Executors.newSingleThreadScheduledExecutor(worker -> new Thread(worker, KidMapper.class.getName()))` since we'll get a single thread we don't need a counter or anything fancy.

it does not help the code flow but if you do a jstack/kill -3/... on the jvm you will identify  the thread instead of having a banalised `pool-1-thread-1` or anything like that.

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: I think, I got the thread-pool handling wrong, so I will rework.

File: src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java
Comment (ArneLimburg) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: OK, and what if the client has no executor set?
Should we use our ScheduledExecutorService then?

Comment (rmannibucau) [reply src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java]: ok for me yes

**Comment by ArneLimburg** [comment]
@rmannibucau Would you please take a look at the changes?

**Review by rmannibucau — APPROVED** [review]

**Comment by ArneLimburg** [comment]
@rmannibucau Thank you for reviewing. What is the process to get this merged and get a release out?
Note: I have no write access to this repository

**Comment by rmannibucau** [comment]
@ArneLimburg my idea was 1. let it open the week end to let a chance to others to review, 2. get it merged next week - can do it, 3. target a release the week after if somebody is available (or the week after if I need to do it - have some constraints before). But happy if anyone wants to go faster too.

**Review by fpapon — APPROVED** [review]

**Pull request merged**
